### PR TITLE
[TD] New command: Surface finish symbols

### PR DESCRIPTION
**Pull request by edi271**
Introducing new command "Surface finish symbols" as discussed in https://forum.freecadweb.org/viewtopic.php?f=35&t=66071.

Description in wiki will follow in the next days.

@WandererFan: please delete commit f16562a, i do not know why it is in this PR.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [X]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [X]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [X]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [X]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [X]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7220">pipeline status</a> for feature branch [PR_7220](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7220). Pipeline [591357706](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/591357706) was triggered at [dda5699](https://github.com/FreeCAD/FreeCAD/pull/7220/commits/dda5699e442d365b4abdff10cc27aae6eea0515d). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/TechDraw/Gui/CommandAnnotate.cpp [hunk 1]
```diff
@@ -1460,6 +1461,50 @@ bool CmdTechDrawWeldSymbol::isActive(void)
...
+{
+    Q_UNUSED(iMsg);
+    std::vector<Gui::SelectionObject> selection = this->getSelection().getSelectionEx();
+    if (selection.empty())
+    {
+        QMessageBox::warning(Gui::getMainWindow(),QObject::tr("SurfaceFinishSymbols"),QObject::tr("Selection is empty"));
+        return;
+    }
+    TechDraw::DrawViewPart* objFeat = dynamic_cast<TechDraw::DrawViewPart*> (selection[0].getObject());
+    if(objFeat == nullptr)
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/CommandAnnotate.cpp:1494]:
```suggestion
    if(!objFeat)
```

File: src/Mod/TechDraw/Gui/CommandAnnotate.cpp [hunk 2]
```diff
@@ -1460,6 +1461,50 @@ bool CmdTechDrawWeldSymbol::isActive(void)
...
+        return;
+    }
+    Gui::Control().showDialog(new TechDrawGui::TaskDlgSurfaceFinishSymbols(objFeat));
+}
+
+bool CmdTechDrawSurfaceFinishSymbols::isActive(void)
+{
+    bool havePage = DrawGuiUtil::needPage(this);
+    bool haveView = DrawGuiUtil::needView(this);
+    return (havePage && haveView);
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/CommandAnnotate.cpp:1506]:
```suggestion
    return havePage && haveView;
```

File: src/Mod/TechDraw/Gui/TaskSurfaceFinishSymbols.cpp [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+        if (ui->pbIcon05 == pressedButton) activeIcon = removeProhibitAll;
+        if (ui->pbIcon06 == pressedButton) activeIcon = removeRequiredAll;
+        
+        QIcon symbolIcon = pressedButton->icon();
+        QGraphicsPixmapItem* pixmapItem = new(QGraphicsPixmapItem);
+        pixmapItem->setPixmap(symbolIcon.pixmap(50,64));
+        pixmapItem->setPos(-50,-130);
+        pixmapItem->setZValue(-1);
+        symbolScene->addItem(pixmapItem);
+    }
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/TaskSurfaceFinishSymbols.cpp:341]:
```suggestion
    QObject* senderObj(this->sender());
    QPushButton* pressedButton = qobject_cast<QPushButton*>(senderObj);
    if (!pressedButton) {
        return;
    }

    if (ui->pbIcon01 == pressedButton) activeIcon = anyMethod;
    if (ui->pbIcon02 == pressedButton) activeIcon = removeProhibit;
    if (ui->pbIcon03 == pressedButton) activeIcon = removeRequired;
    if (ui->pbIcon04 == pressedButton) activeIcon = anyMethodAll;
    if (ui->pbIcon05 == pressedButton) activeIcon = removeProhibitAll;
    if (ui->pbIcon06 == pressedButton) activeIcon = removeRequiredAll;
    
    QIcon symbolIcon = pressedButton->icon();
    QGraphicsPixmapItem* pixmapItem = new(QGraphicsPixmapItem);
    pixmapItem->setPixmap(symbolIcon.pixmap(50,64));
    pixmapItem->setPos(-50,-130);
    pixmapItem->setZValue(-1);
    symbolScene->addItem(pixmapItem);
```

File: src/Mod/TechDraw/Gui/TaskSurfaceFinishSymbols.cpp [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+#include <QGraphicsScene>
+#include <QLineEdit>
+#include <QGraphicsProxyWidget>
+#include <QComboBox>
+#include <QString>
+#include <QXmlStreamReader>
+#include <QByteArray>
+#include <QSvgRenderer>
+#include <QColor>
+#include <QGraphicsPixmapItem>
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/TaskSurfaceFinishSymbols.cpp:81]: Are all this really used?!

**Review by benj5378 — COMMENTED** [review]
Details

File: src/Mod/TechDraw/Gui/TaskSurfaceFinishSymbols.cpp (hunk 4 above)
Comment (edi271) [reply src/Mod/TechDraw/Gui/TaskSurfaceFinishSymbols.cpp]: See bc6ed8e

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7220">pipeline status</a> for feature branch [PR_7220](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7220). Pipeline [592047313](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/592047313) was triggered at [bc6ed8e](https://github.com/FreeCAD/FreeCAD/pull/7220/commits/bc6ed8e4bffd4d5b6c3215254f9dd07ab1ef5ffb). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7220">pipeline status</a> for feature branch [PR_7220](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7220). Pipeline [592058411](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/592058411) was triggered at [384b7c3](https://github.com/FreeCAD/FreeCAD/pull/7220/commits/384b7c322ce648a848e4321a3df0841c346d883e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by WandererFan** [comment]
minor items: 
- average roughness drop down is obscured by lay symbol combo box or scrolls out of display area
- decimal separator in RA should match locale
- backgrounds/shadows on widgets don't add anything
- is the symbol supposed to be attached to the view?

**Comment by WandererFan** [comment]
AreaAnnotation icon replaced, rebased to master and commits squashed in #7227.

**Pull request closed without merging**